# Lab1 - Scikit-learn
Author: *YOUR NAME*

## 1. Introduction

The goal of this lab is to become familiar with the scikit-learn library.

You will practice loading example datasets, perform classification and regression with linear scikit-learn models, and investigate the effects of reducing the number of features (columns in X) and the number of samples (rows in X and y).


In [179]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

## 2. Classification

Using yellowbrick spam - classification  
https://www.scikit-yb.org/en/latest/api/datasets/spam.html

The goal is to investigate `LogisticRegression(max_iter=2000)` and effects of reducing the number of features and number of samples on classification performance.

### 2.1 Implement convenience function

In [180]:
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split

def get_classifier_accuracy(model, X, y):
    '''Calculate train and validation accuracy of classifier (model)
        
        Splits feature matrix X and target vector y 
        with sklearn train_test_split() and random_state=956.
        
        model (sklearn classifier): Classifier to train and evaluate
        X (numpy.array or pandas.DataFrame): Feature matrix
        y (numpy.array or pandas.Series): Target vector
        
        returns: training accuracy, validation accuracy
    
    '''


    
    #TODO: IMPLEMENT FUNCTION BODY
    
    X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=956)
    model.fit(X_train, y_train)
    y_pred_train = model.predict(X_train)   
    y_pred_test = model.predict(X_test)
    train_acc = accuracy_score(y_train, y_pred_train)
    test_acc = accuracy_score(y_test, y_pred_test)
    return train_acc, test_acc

### 2.2 Load data

Use the yellowbrick function `load_spam()`, load the spam data set into feature matrix `X` and target vector `y`.

Print size and type of `X` and `y`.


In [181]:
# TODO: ADD YOUR CODE HERE



from yellowbrick import datasets

X, y = datasets.load_spam() 
print(X.shape)
print(type(X))
print(y.shape)
print(type(y))

X.head()




(4600, 57)
<class 'pandas.core.frame.DataFrame'>
(4600,)
<class 'pandas.core.series.Series'>


word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0            0.21               0.28           0.50           0.0   
1            0.06               0.00           0.71           0.0   
2            0.00               0.00           0.00           0.0   
3            0.00               0.00           0.00           0.0   
4            0.00               0.00           0.00           0.0   

   word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0           0.14            0.28              0.21                0.07   
1           1.23            0.19              0.19                0.12   
2           0.63            0.00              0.31                0.63   
3           0.63            0.00              0.31                0.63   
4           1.85            0.00              0.00                1.85   

   word_freq_order  word_freq_mail  ...  word_freq_conference  char_freq_;  \
0             0.00            0.94  ...                   0.0         0.00   
1             0.64            0.25  ...                   0.0         0.01   
2             0.31            0.63  ...                   0.0         0.00   
3             0.31            0.63  ...                   0.0         0.00   
4             0.00            0.00  ...                   0.0         0.00   

   char_freq_(  char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
0        0.132          0.0        0.372        0.180        0.048   
1        0.143          0.0        0.276        0.184        0.010   
2        0.137          0.0        0.137        0.000        0.000   
3        0.135          0.0        0.135        0.000        0.000   
4        0.223          0.0        0.000        0.000        0.000   

   capital_run_length_average  capital_run_length_longest  \
0                       5.114                         101   
1                       9.821                         485   
2                       3.537                          40   
3                       3.537                          40   
4                       3.000                          15   

   capital_run_length_total  
0                      1028  
1                      2259  
2                       191  
3                       191  
4                        54  

[5 rows x 57 columns]

Using the sklearn function `train_test_split()` prepare a feature matrix `X_small` and target vector `y_small` that contain only **1%** of the rows. Use `random_state=174`.

Print size and type of `X_small` and `y_small`.

In [182]:
# TODO: ADD YOUR CODE HERE



X_small, _, y_small, _ = train_test_split(X, y, train_size=0.01, random_state=174)
print(X_small.shape)
print(type(X_small))
print(y_small.shape)
print(type(y_small))


(46, 57)
<class 'pandas.core.frame.DataFrame'>
(46,)
<class 'pandas.core.series.Series'>


### 2.3 Train and evaluate models

1. Import `LogisticRegression` from sklearn
2. Instantiate model `LogisticRegression(max_iter=2000)`.
3. Create a pandas DataFrame `results` with columns: Data size, training accuracy, validation accuracy
4. Call your convenience function `get_classifier_accuracy()` using 
    - `X` and `y`
    - Only first two columns of `X` and `y`
    - `X_small` and `y_small`
5. Add the data size, training and validation accuracy for each call to the `results` DataFrame
6. Print `results`

In [183]:
# TODO: ADD YOUR CODE HERE
from sklearn.linear_model import LogisticRegression


model = LogisticRegression(max_iter=2000)

results = pd.DataFrame(columns=["data size", "training acc", "validation acc"])


In [184]:
#print(get_classifier_accuracy(model, X,y))
# get_classifier_accuracy(only first two columns of X,y)


results = results.append({"data size": X.shape, "training acc": get_classifier_accuracy(model, X,y)[0], "validation acc": get_classifier_accuracy(model, X,y)[1]}, ignore_index=True)

results  = results.append({"data size": X.iloc[:,:2].shape, "training acc": get_classifier_accuracy(model, X.iloc[:,:2],y)[0], "validation acc": get_classifier_accuracy(model, X.iloc[:,:2],y)[1]}, ignore_index=True)

results = results.append({"data size": X_small.shape, "training acc": get_classifier_accuracy(model, X_small,y_small)[0], "validation acc": get_classifier_accuracy(model, X_small,y_small)[1]}, ignore_index=True)


# print(get_classifier_accuracy(model, X.iloc[:,:2],y))

# print(get_classifier_accuracy(model, X_small,y_small))


/var/folders/jb/h7p_6zx90lq3mlgz6102t6ww0000gn/T/ipykernel_93608/798509156.py:5: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  results = results.append({"data size": X.shape, "training acc": get_classifier_accuracy(model, X,y)[0], "validation acc": get_classifier_accuracy(model, X,y)[1]}, ignore_index=True)
/var/folders/jb/h7p_6zx90lq3mlgz6102t6ww0000gn/T/ipykernel_93608/798509156.py:7: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  results  = results.append({"data size": X.iloc[:,:2].shape, "training acc": get_classifier_accuracy(model, X.iloc[:,:2],y)[0], "validation acc": get_classifier_accuracy(model, X.iloc[:,:2],y)[1]}, ignore_index=True)
/var/folders/jb/h7p_6zx90lq3mlgz6102t6ww0000gn/T/ipykernel_93608/798509156.py:9: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a

In [185]:
results

data size training acc validation acc
0  (4600, 57)     0.935072       0.917391
1   (4600, 2)     0.608986       0.613043
2    (46, 57)     0.941176           0.75

### 2.4 Questions
1. What is the validation accuracy using all data? What is the difference between training and validation accuracy?
1. How does the validation accuracy and difference between training and validation change when only two columns are used? Provide values.
1. How does the validation accuracy and difference between training and validation change when only 1% of the rows are used? Provide values.

*YOUR ANSWERS HERE*




1. Using all the data: 91.73% accuracy. Training accuracy measures how accurately our model fits the data it was trained on. Validation accuracy measures how accurately our model fits new unseen data, which it was not trained on.

2. Comparing all columns with only two columns, the validation accuracy decreases from 91.73% to 61.30 %. The Training accuracy decreases from 93.5% to 60.8%. 

3. Comparing using all rows vs only 1% of rows, the Training accuracy increased from 93.5% to 94.11% , the validation accuracy decreased from 91.73% to 75% 

## 3. Regression

Using yellowbrick energy - regression  
https://www.scikit-yb.org/en/latest/api/datasets/energy.html

The goal is to investigate `LinearRegression()` and effects of reducing the number of features and number of samples on regression performance.

### 3.1 Implement convenience function

In [186]:
from sklearn.metrics import mean_squared_error
from sklearn.model_selection import train_test_split

def get_regressor_mse(model, X, y):
    '''Calculate train and validation mean-squared error (mse) of regressor (model)
        
        Splits feature matrix X and target vector y 
        with sklearn train_test_split() and random_state=956.
        
        model (sklearn regressor): Regressor to train and evaluate
        X (numpy.array or pandas.DataFrame): Feature matrix
        y (numpy.array or pandas.Series): Target vector
        
        returns: training mse, validation mse
    
    '''
   
    #TODO: IMPLEMENT FUNCTION BODY
    X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=956)
    model.fit(X_train, y_train)
    y_pred_train = model.predict(X_train)
    y_pred_test = model.predict(X_test)
    train_mse = mean_squared_error(y_train, y_pred_train)
    test_mse = mean_squared_error(y_test, y_pred_test)
    return train_mse, test_mse
    

### 3.2 Load data

Use the yellowbrick function `load_energy()` load the energy data set into feature matrix `X` and target vector `y`.

Print dimensions and type of `X` and `y`.

In [187]:
# TODO: ADD YOUR CODE HERE
Xe,ye=datasets.load_energy()

print(Xe.shape)
print(type(Xe))
print(ye.shape)
print(type(ye))


(768, 8)
<class 'pandas.core.frame.DataFrame'>
(768,)
<class 'pandas.core.series.Series'>


Using the sklearn function `train_test_split()` prepare a feature matrix `X_small` and target vector `y_small` that contain only **1%** of the rows. Use `random_state=174`.

Print size and type of `X_small` and `y_small`.

In [188]:
# TODO: ADD YOUR CODE HERE
Xe_small, _, ye_small, _ = train_test_split(Xe, ye, train_size=0.01, random_state=174)

print(Xe_small.shape)
print(type(Xe_small))
print(ye_small.shape)
print(type(ye_small))



(7, 8)
<class 'pandas.core.frame.DataFrame'>
(7,)
<class 'pandas.core.series.Series'>


### 3.3 Train and evaluate models

1. Import `LinearRegression` from sklearn
2. Instantiate model `LinearRegression()`.
3. Create a pandas DataFrame `results` with columns: Data size, training MSE, validation MSE
4. Call your convenience function `get_regressor_mse()` using 
    - `X` and `y`
    - Only first two columns of `X` and `y`
    - `X_small` and `y_small`
5. Add the data size, training and validation MSE for each call to the `results` DataFrame
6. Print `results`

In [189]:
# TODO: ADD YOUR CODE HERE
from sklearn.linear_model import LinearRegression

model_LR = LinearRegression()

results_LR = pd.DataFrame(columns=["data size", "training mse", "validation mse"])

results_LR = results_LR.append({"data size": Xe.shape, "training mse": get_regressor_mse(model_LR, Xe,ye)[0], "validation mse": get_regressor_mse(model_LR, Xe,ye)[1]}, ignore_index=True)

results_LR = results_LR.append({"data size": Xe.iloc[:,:2].shape, "training mse": get_regressor_mse(model_LR, Xe.iloc[:,:2],ye)[0], "validation mse": get_regressor_mse(model_LR, Xe.iloc[:,:2],ye)[1]}, ignore_index=True)

results_LR = results_LR.append({"data size": Xe_small.shape, "training mse": get_regressor_mse(model_LR, Xe_small,ye_small)[0], "validation mse": get_regressor_mse(model_LR, Xe_small,ye_small)[1]}, ignore_index=True)
# print(get_regressor_mse(model_LR, Xe,ye))
# print(get_regressor_mse(model_LR, Xe.iloc[:,:2],ye))
# print(get_regressor_mse(model_LR, Xe_small,ye_small))



/var/folders/jb/h7p_6zx90lq3mlgz6102t6ww0000gn/T/ipykernel_93608/3555247351.py:8: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  results_LR = results_LR.append({"data size": Xe.shape, "training mse": get_regressor_mse(model_LR, Xe,ye)[0], "validation mse": get_regressor_mse(model_LR, Xe,ye)[1]}, ignore_index=True)
/var/folders/jb/h7p_6zx90lq3mlgz6102t6ww0000gn/T/ipykernel_93608/3555247351.py:10: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  results_LR = results_LR.append({"data size": Xe.iloc[:,:2].shape, "training mse": get_regressor_mse(model_LR, Xe.iloc[:,:2],ye)[0], "validation mse": get_regressor_mse(model_LR, Xe.iloc[:,:2],ye)[1]}, ignore_index=True)
/var/folders/jb/h7p_6zx90lq3mlgz6102t6ww0000gn/T/ipykernel_93608/3555247351.py:12: FutureWarning: The frame.append method is deprecated and will be removed fr

In [190]:
results_LR

data size training mse validation mse
0  (768, 8)     7.969557      10.323016
1  (768, 2)     53.60043      46.410426
2    (7, 8)          0.0      69.977449

### 3.4 Questions
1. What is the validation MSE using all data? What is the difference between training and validation MSE?
1. How does the validation MSE and difference between training and validation change when only two columns are used? Provide values.
1. How does the validation MSE and difference between training and validation change when only 1% of the rows are used? Provide values.

*YOUR ANSWERS HERE*



1. Using all the data, Training MSE is 7.97. Training MSE measures how accurately our model fits the data it was trained on. Validation MSE using all the data is 10.32 Validation MSE measures how accurately our model fits new unseen data, which it was not trained on.

2. Training MSE increases from 7.97 to 53.6 Validation MSE increases from 10.32 to 46.41

3. Training MSE decreases from 7.97 to 0.0 Validation MSE increases from 10.32 to 69.98

## 4. Observations/Interpretation

Describe any pattern you see in the results. Relate your findings to what we discussed during lectures. Include data to justify your findings.


*ADD YOUR FINDINGS HERE*



The training accuracy and MSE both decrease when only using 1% of rows, however the validation accuracy and MSE both increase. This is because using only 1% is overfitting the data. New data is not being predicted well because the model is too specific to the small amount of training data. 

When using only first two columns of training data ,  the training/validation accuracy decreases and MSE increases, which is expected because the model is not as accurate as when using all the data to be trained on. 


When using all the data, we get relatively acceptable results. 



## 5. Reflection
Include a sentence or two about:
- what you liked or disliked,
- found interesting, confusing, challangeing, motivating
while working on this assignment.


*ADD YOUR THOUGHTS HERE*



I liked implementing the convenience functions to streamline generating results. Seeing how different results were produced by adjusting the data was interesting. 